pkg/firewall/firewall_test.go

```diff
@@ -0,0 +1,228 @@
+	policy := &stakeOrActiveKeepPolicy{
+		chain:                  chain,
+		minimumStakePolicy:     coreFirewall,
+		activeKeepMembersCache: newTimeCache(cacheLifeTime),
+	}
+
+	_, remotePeerPublicKey, err := key.GenerateStaticNetworkKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	chain.OpenKeep(
+		common.HexToAddress("0xD6e148Be1E36Fc4Be9FE5a1abD7b3103ED527256"),
+		[]common.Address{
+			common.HexToAddress("0x4f7C771Ab173bEc2BbE980497111866383a21172"),
+			common.HexToAddress(key.NetworkPubKeyToEthAddress(remotePeerPublicKey)),
+		},
+	)
+
+	if err := policy.Validate(
+		key.NetworkKeyToECDSAKey(remotePeerPublicKey),
+	); err != nil {
+		t.Fatalf("validation should pass: [%v]", err)
+	}
+}
+
+func TestNoMinimumStakeIsClosedKeepMember(t *testing.T) {
+	chain := local.Connect()
+	coreFirewall := newMockCoreFirewall()
+	policy := &stakeOrActiveKeepPolicy{
+		chain:                  chain,
+		minimumStakePolicy:     coreFirewall,
+		activeKeepMembersCache: newTimeCache(cacheLifeTime),
+	}
+
+	_, remotePeerPublicKey, err := key.GenerateStaticNetworkKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	keepAddress := common.HexToAddress("0xD6e148Be1E36Fc4Be9FE5a1abD7b3103ED527256")
+	chain.OpenKeep(
+		keepAddress,
+		[]common.Address{
+			common.HexToAddress("0x4f7C771Ab173bEc2BbE980497111866383a21172"),
+			common.HexToAddress(key.NetworkPubKeyToEthAddress(remotePeerPublicKey)),
+		},
+	)
+	if err := chain.CloseKeep(keepAddress); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := policy.Validate(
+		key.NetworkKeyToECDSAKey(remotePeerPublicKey),
+	); err != errNoMinStakeNoActiveKeep {
+		t.Fatalf(
+			"unexpected validation error\nactual:   [%v]\nexpected: [%v]",
+			err,
+			errNoMinStakeNoActiveKeep,
+		)
+	}
+}
+
+func TestNoMinimumStakeMultipleKeepsMember(t *testing.T) {
+	chain := local.Connect()
+	coreFirewall := newMockCoreFirewall()
+	policy := &stakeOrActiveKeepPolicy{
+		chain:                  chain,
+		minimumStakePolicy:     coreFirewall,
+		activeKeepMembersCache: newTimeCache(cacheLifeTime),
+	}
+
+	_, remotePeerPublicKey, err := key.GenerateStaticNetworkKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	keep1Address := common.HexToAddress("0xD6e148Be1E36Fc4Be9FE5a1abD7b3103ED527256")
+	chain.OpenKeep(
+		keep1Address,
+		[]common.Address{
+			common.HexToAddress("0x4f7C771Ab173bEc2BbE980497111866383a21172"),
+			common.HexToAddress(key.NetworkPubKeyToEthAddress(remotePeerPublicKey)),
+		},
+	)
+	keep2Address := common.HexToAddress("0x1Ca1EB1CafF6B3784Fe28a1b12266a10D04626A0")
+	chain.OpenKeep(
+		keep2Address,
+		[]common.Address{
+			common.HexToAddress("0xF9798F39CfEf21931d3B5F73aF67718ae569a73e"),
+			common.HexToAddress(key.NetworkPubKeyToEthAddress(remotePeerPublicKey)),
+		},
+	)
+
+	if err := chain.CloseKeep(keep1Address); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := policy.Validate(
+		key.NetworkKeyToECDSAKey(remotePeerPublicKey),
+	); err != nil {
+		t.Fatalf("validation should pass: [%v]", err)
+	}
+}
+
```

Comment:
Can we add a case where we validate a member that already exists in the cache?

Reply:
I added a test case ensuring we cache active keep members but it's hard to write a good test ensuring we don't hit the chain in case the given member is already in the cache. This would be possible if we had a separate chain interface for the factory, similarly how we have a separate chain interface for signing or threshold relay or groups in `keep-core`. Without this refactoring, it's hard to implement such a case without dirty hacks. Or at least I have no idea how to do it without dirty hacks.

Reply:
Do we want to create an issue to do that in the future?

Reply:
https://github.com/keep-network/keep-ecdsa/issues/420